src/proto/h1/io.rs

```diff
@@ -164,14 +164,23 @@ where
         self.read_blocked
     }
 
-    pub fn flush(&mut self) -> Poll<(), io::Error> {
+    pub fn flush(&mut self, cx: &mut task::Context) -> Poll<(), io::Error> {
         if self.flush_pipeline && !self.read_buf.is_empty() {
             //Ok(())
         } else if self.write_buf.remaining() == 0 {
-            try_nb!(self.io.flush());
+            try_ready!(self.io.poll_flush(cx));
         } else {
             loop {
-                let n = try_ready!(self.io.write_buf(&mut self.write_buf.auto()));
+                let (n, strategy) = try_ready!(write_buf(&mut self.io, cx, &mut self.write_buf));
+                if self.write_buf.strategy == Strategy::Auto {
```

Comment:
I think this can still be done with `WriteBufAuto`, that allows only doing the strategy checking as long as it's still `Auto`.

Reply:
Can you go into more detail about that? Why would we need `WriteBufAuto`?

Reply:
I mostly was thinking of these reasons:

1. Allows the strategy swapping code to be separate from this, which is just wanting to write until blocked.
2. The checks to determine what strategy to switch to could probably only have to happen when strategy it still auto.

Reply:
Ah, cool! I can definitely factor out the strategy swapping code, I guess `WriteBufAuto` would now implement some sort of `write_into` method that takes a `AsyncWrite`?

- [ ] factor out strategy swapping code